atomicfu/src/nativeMain/kotlin/kotlinx/atomicfu/Synchronized.kt

```diff
@@ -0,0 +1,204 @@
+
+import platform.posix.*
+import interop.*
+import kotlinx.cinterop.*
+import kotlin.native.concurrent.AtomicNativePtr
+import kotlin.native.concurrent.AtomicReference
+import kotlin.native.concurrent.SharedImmutable
+import kotlin.native.concurrent.freeze
+import kotlin.native.internal.NativePtr
+import kotlinx.atomicfu.Status.*
+
+public actual open class SynchronizedObject {
+
+    private val lock = AtomicReference(LockState(UNLOCKED, 0, 0))
+
+    public fun lock() {
+        val currentThreadId = pthread_self()
+        while (true) {
+            val state = lock.value
+            when (state.status) {
+                UNLOCKED -> {
+                    val thinLock = LockState(THIN, 1, 0, currentThreadId)
+                    if (lock.compareAndSet(state, thinLock)) return
+                }
+                THIN -> {
+                    if (currentThreadId == state.ownerThreadId) {
+                        // reentrant lock
+                        val thinNested = LockState(THIN, state.nestedLocks + 1, state.waiters, currentThreadId)
+                        if (lock.compareAndSet(state, thinNested)) return
+                    } else {
+                        // another thread is trying to take this lock -> allocate native mutex
+                        val mutex = mutexPool.allocate(locked = true)
+                        val fatLock = LockState(FAT, state.nestedLocks, state.waiters + 1, state.ownerThreadId, mutex)
+                        if (lock.compareAndSet(state, fatLock)) {
+                            //block the current thread waiting for the owner thread to release the permit
+                            mutex.lock()
+                            tryLockAfterResume(currentThreadId)
+                            return
+                        } else {
+                            // return permit taken for the owner thread and release mutex back to the pool
+                            mutex.unlock()
+                            mutexPool.release(mutex)
+                        }
+                    }
+                }
+                FAT -> {
+                    if (currentThreadId == state.ownerThreadId) {
+                        // nested lock
+                        val nestedFatLock = LockState(FAT, state.nestedLocks + 1, state.waiters, state.ownerThreadId, state.mutex)
+                        if (lock.compareAndSet(state, nestedFatLock)) return
+                    } else {
+                        val fatLock = LockState(FAT, state.nestedLocks, state.waiters + 1, state.ownerThreadId, state.mutex)
+                        if (lock.compareAndSet(state, fatLock)) {
+                            fatLock.mutex!!.lock()
+                            tryLockAfterResume(currentThreadId)
+                            return
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public fun unlock() {
+        val currentThreadId = pthread_self()
```

Comment:
This does not compile on MacOS. You need `pthread_self()!!` (it is nullable there). There will be warning during Linux compilation, though. You can avoid it by doing a custom wrapper.